language/move-vm/runtime/src/runtime.rs

```diff
@@ -111,58 +130,126 @@ impl VMRuntime {
+            if n_signer_params != n_signers {
+                return Err(
+                    PartialVMError::new(StatusCode::NUMBER_OF_SIGNER_ARGUMENTS_MISMATCH)
+                        .with_message(format!(
+                            "Expected {} signer args got {}",
+                            n_signer_params, n_signers
+                        )),
+                );
+            }
+            let mut vals: Vec<Value> = senders
+                .into_iter()
+                .map(Value::transaction_argument_signer_reference)
+                .collect();
+            vals.extend(self.deserialize_args(&tys[n_signers..], args)?);
+            vals
+        };
+
+        Ok(args)
+    }
+
     // See Session::execute_script for what contracts to follow.
     pub(crate) fn execute_script(
         &self,
         script: Vec<u8>,
         ty_args: Vec<TypeTag>,
-        mut args: Vec<Value>,
+        args: Vec<Vec<u8>>,
         senders: Vec<AccountAddress>,
         data_store: &mut impl DataStore,
         cost_strategy: &mut CostStrategy,
         log_context: &impl LogContext,
     ) -> VMResult<()> {
-        // signer helper closure
-        fn is_signer_reference(s: &SignatureToken) -> bool {
-            use SignatureToken as S;
-            match s {
-                S::Reference(inner) => matches!(&**inner, S::Signer),
-                _ => false,
-            }
-        }
-
         // load the script, perform verification
-        let (main, type_params) =
+        let (main, ty_args, params) =
             self.loader
                 .load_script(&script, &ty_args, data_store, log_context)?;
 
-        // Build the arguments list for the main and check the arguments are of restricted types.
-        // Signers are built up from left-to-right. Either all signer arguments are used, or no
-        // signer arguments can be be used by a script.
-        let parameters = &main.parameters().0;
-        let has_signer_parameters = parameters.get(0).map_or(false, is_signer_reference);
-        let mut signers_and_args = if has_signer_parameters {
-            if parameters.len() != args.len() + senders.len() {
-                return Err(PartialVMError::new(StatusCode::TYPE_MISMATCH)
-                    .with_message("Scripts must use all or no signers".to_string())
-                    .finish(Location::Script));
-            }
-            // add signers to args
-            senders
-                .into_iter()
-                .map(Value::transaction_argument_signer_reference)
-                .collect()
-        } else {
-            // no signer parameters, don't add to args
-            vec![]
-        };
-        signers_and_args.append(&mut args);
-        check_args(&signers_and_args).map_err(|e| e.finish(Location::Script))?;
-
+        let signers_and_args = self
+            .create_signers_and_arguments(&params, senders, args)
+            .map_err(|err| err.finish(Location::Undefined))?;
```

Comment:
```suggestion
            .map_err(|err| err.finish(Location::Script))?;
```
or perhaps read the start location from `main`?

Reply:
Its an interesting one. 
Typically "Undefined" is used for anything that is VM specific and doesn't really live in any location. For example if you hit an error in the loader, you get Undefined even if that was the result of loading some module. 

So here I could see it going eitherway, since its for the script args, but its not really *in* the script since the script has not started executing yet

Reply:
Yeah, good point. This is really like a bytecode verifier error; keeping locations for errors that occur during concrete execution seems very sensible.

Reply:
Shall we consider adding something like a `Location::TransactionArgument`? To me that's also a common source of user errors.

Reply:
> Shall we consider adding something like a Location::TransactionArgument? 
I feel like that bucket is a bit narrow, but maybe

> Yeah, good point. This is really like a bytecode verifier error; keeping locations for errors that occur during concrete execution seems very sensible.
From a bytecode verifier perspective, this should probably be Script, since the location is set for the verifier. 
But I think it depends on what this error really is semantically. 

So script might make sense here